Sources/CombineGRPC/Server/BidirectionalStreamingHandlerSubscriber.swift

```diff
@@ -32,8 +32,15 @@ class BidirectionalStreamingHandlerSubscriber<Request, Response>: Subscriber, Ca
   func receive(completion: Subscribers.Completion<RPCError>) {
     switch completion {
     case .failure(let error):
-      context.trailers = augment(headers: context.trailers, with: error)
-      context.statusPromise.fail(error.status)
+      if context.eventLoop.inEventLoop {
+        context.trailers = augment(headers: context.trailers, with: error)
+        context.statusPromise.fail(error.status)
+      } else {
+        context.eventLoop.execute {
+          self.context.trailers = augment(headers: self.context.trailers, with: error)
+          self.context.statusPromise.fail(error.status)
+        }
+      }
```

Comment:
The above snippet is repeated for each server side handler.

Have you considered removing [the augment(headers:with) global function](https://github.com/vyshane/grpc-swift-combine/blob/master/Sources/CombineGRPC/Server/HeaderUtils.swift#L7) and perhaps replacing it with a `fail(context:with)` function? We then pass in the context itself.

Reply:
I'm not sure exactly what you mean. The only thing I was doing was ensure the assignment to `context.trailer` was always done on the event loop thread.

The augment function can be called on any thread but the assignment needs to be on the correct thread.

Reply:
Oh, I see. You mean refactor this code into a common function. I can do that.

Reply:
Actually the type are not common enough in GRPC. The four calls basically require 3 separate functions.